proposals/lip-0039.md

```diff
@@ -302,273 +328,379 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
+            return False
+            
+    return verify(queryKeys, proof, root)
 ```
 
-```java
-function calculateRoot(siblingHashes, queries):
-    // we add the hash property to each query, used to recompute the tree root
-    for each query q in queries:
-        set q.binaryBitmap to binaryExpansion(q.bitmap) without leading zeros
-        // if value is empty, it is a defaultEmptyNode
-        if q.value is empty bytes:
-            set q.hash to EMPTY_HASH
+```python
+def verify(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
+    for key, query in zip(queryKeys, proof.queries):
+        # The bitmap does not contain extra leading 0s.
+        if query.bitmap[0] == 0:
+            return False
+        if len(key) != KEY_LENGTH_BYTES:
+            return False
+
+        # q is an inclusion proof for key or a default empty node.
+        if key == query.key:
+            continue
+
+        # q is an inclusion proof for another leaf node.
+        commonPrefix = os.path.commonprefix([binaryExpansion(query.key), binaryExpansion(key)])
+        if len(query.binaryBitmap) > len(commonPrefix):
+            # q does not give an non-inclusion proof for key.
+            return False
+
+    filteredQueries = []
+    _filter: dict[bytes, Query] = {}
+    for q in proof.queries:
+        # Remove duplicate queries preserving order.
+        # This can happen if the same query is given for different queryKeys, as for inclusion proofs or non-inclusion proofs pointing to a leaf node
+        # or for non-inclusion proofs pointing to the same empty node.
+        # To do this, we check the binary path (key binary expansion up to the query height).
+        if q.binaryPath not in _filter:
+            filteredQueries.append(q)
+            _filter[q.binaryPath] = q
+        # If the query is a duplicate, check that the two queries are consistent.
         else:
-            let qLeaf=leafNode(q.key,q.value)
-            set q.hash to qLeaf.hash
-
-    let sortedQueries=sortByBitmapAndKey(queries)
-    sortedQueries=filterQueries(sortedQueries)
-    while sortedQueries is non-empty:
-        let q be the first element of sortedQueries
-        // if the binaryBitmap is empty string, we reached the top of the tree
-        if q.binaryBitmap is empty string:
+            duplicateQuery = _filter[q.binaryPath]
+            # Check that the value and bitmap are the same.
+            if q.value != duplicateQuery.value or q.bitmap != duplicateQuery.bitmap:
+                return False
+            # If the value is empty, both queries point to an empty node.
+            # The binary paths must match (the keys can be different).
+            if q.value == EMPTY_VALUE:
+                if q.binaryPath != duplicateQuery.binaryPath:
+                    return False
+            # Otherwise, the keys must be the same.
+            else:
+                if q.key != duplicateQuery.key:
+                    return False
```

Comment:
I joined the checks

Reply:
Maybe the comment could still be updated to
"In case of inclusion proofs, i.e., if the value of q is empty, the keys must be the same".
Do we here implicitly assume to only either exclusively have inclusion or exclusive non-inclusion proofs?

Reply:
we don't assume that, they can be mixed.

Reply:
Regarding the comment, in case of inclusion proof the value of q is non empty. Also, if the value of q is non empty, this is not necessarily an inclusion proof